pkg/store/proxy.go

```diff
@@ -255,305 +270,209 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
-		return frameTimeoutCtx, cancel
-	}
-	return frameTimeoutCtx, func() {}
-}
+				if rr.err != nil {
+					err := errors.Wrapf(rr.err, "receive series from %s", st.String())
+					errMtx.Lock()
+					errs.Add(err)
+					errMtx.Unlock()
 
-func startStreamSeriesSet(
-	ctx context.Context,
-	logger log.Logger,
-	span tracing.Span,
-	closeSeries context.CancelFunc,
-	wg *sync.WaitGroup,
-	stream storepb.Store_SeriesClient,
-	warnCh directSender,
-	name string,
-	partialResponse bool,
-	responseTimeout time.Duration,
-	emptyStreamResponses prometheus.Counter,
-) *streamSeriesSet {
-	s := &streamSeriesSet{
-		ctx:             ctx,
-		logger:          logger,
-		closeSeries:     closeSeries,
-		stream:          stream,
-		warnCh:          warnCh,
-		recvCh:          make(chan *storepb.Series, 10),
-		name:            name,
-		partialResponse: partialResponse,
-		responseTimeout: responseTimeout,
-	}
+					span.SetTag("err", err.Error())
+					span.Finish()
+					close(done)
+					return false
+				}
 
-	wg.Add(1)
-	go func() {
-		seriesStats := &storepb.SeriesStatsCounter{}
-		bytesProcessed := 0
-
-		defer func() {
-			span.SetTag("processed.series", seriesStats.Series)
-			span.SetTag("processed.chunks", seriesStats.Chunks)
-			span.SetTag("processed.samples", seriesStats.Samples)
-			span.SetTag("processed.bytes", bytesProcessed)
-			span.Finish()
-			close(s.recvCh)
-			wg.Done()
-		}()
-
-		numResponses := 0
-		defer func() {
-			if numResponses == 0 {
-				emptyStreamResponses.Inc()
+				storeResponses[i].responses = append(storeResponses[i].responses, rr.r)
+				return true
 			}
-		}()
 
-		rCh := make(chan *recvResponse)
-		done := make(chan struct{})
-		go func() {
 			for {
-				r, err := s.stream.Recv()
-				select {
-				case <-done:
-					close(rCh)
+				if !handleRecvResponse() {
 					return
-				case rCh <- &recvResponse{r: r, err: err}:
 				}
 			}
-		}()
-		// The `defer` only executed when function return, we do `defer cancel` in for loop,
-		// so make the loop body as a function, release timers created by context as early.
-		handleRecvResponse := func() (next bool) {
-			frameTimeoutCtx, cancel := frameCtx(s.responseTimeout)
-			defer cancel()
-			var rr *recvResponse
-			select {
-			case <-ctx.Done():
-				s.handleErr(errors.Wrapf(ctx.Err(), "failed to receive any data from %s", s.name), done)
-				return false
-			case <-frameTimeoutCtx.Done():
-				s.handleErr(errors.Wrapf(frameTimeoutCtx.Err(), "failed to receive any data in %s from %s", s.responseTimeout.String(), s.name), done)
-				return false
-			case rr = <-rCh:
-			}
+		}(i, st)
+	}
 
-			if rr.err == io.EOF {
-				close(done)
-				return false
-			}
+	level.Debug(reqLogger).Log("msg", "Series: started fanout streams", "status", strings.Join(storeDebugMsgs, ";"))
 
-			if rr.err != nil {
-				s.handleErr(errors.Wrapf(rr.err, "receive series from %s", s.name), done)
-				return false
-			}
-			numResponses++
-			bytesProcessed += rr.r.Size()
+	wg.Wait()
```

Comment:
I think this can work for a start, but we can actually start k-merging much sooner - as soon as we have first frames from each store. 🤔 

We can start with this simple solution (which is already better since we buffer ASAP), but let's ensure this code does not need to be refactored full to start merging asynchronously, WDYT?

Reply:
Refactored so that it would be a one-line change to switch to lazy evaluation.